mkosi/machine.py

```diff
@@ -0,0 +1,129 @@
+    MKOSI_COMMANDS_SUDO,
+    build_stuff,
+    bump_image_version,
+    check_native,
+    check_output,
+    check_root,
+    init_namespace,
+    load_args,
+    needs_build,
+    parse_args,
+    prepend_to_environ_path,
+    print_output_size,
+    run_qemu_cmdline,
+    run_shell_cmdline,
+    save_manifest,
+    suppress_stacktrace,
+    unlink_output,
+    virt_name,
+)
+from .backend import die, run
+
+
+class Machine:
+    def __init__(self, args: Optional[Sequence[str]] = None) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self._serial: Optional[pexpect.spawnu] = None
+        self.stack = contextlib.ExitStack()
+        self.args = load_args(parse_args(args)["default"])
+        self.verb = self.args.verb
+        self.exit_code: int
+
+    @property
+    def serial(self) -> pexpect.spawnu:
+        if self._serial is None:
+            raise ValueError("Trying to access serial console before machine boot or after machine shutdown")
+
+        return self._serial
+
+    @serial.setter
+    def serial(self, cmdline: List[str]) -> pexpect.spawnu:
+        self._serial = pexpect.spawnu(cmdline, logfile=None, timeout=240)
+
+        return self._serial
+
+    def __enter__(self) -> Machine:
+        with contextlib.ExitStack() as stack:
+            prepend_to_environ_path(self.args.extra_search_paths)
+
+            if self.args.verb in MKOSI_COMMANDS_SUDO:
+                check_root()
+                unlink_output(self.args)
+
+            if self.args.verb == "build":
+                check_output(self.args)
+
+            if needs_build(self.args):
+                check_root()
+                check_native(self.args)
+                init_namespace(self.args)
+                manifest = build_stuff(self.args)
+
+                if self.args.auto_bump:
+                    bump_image_version(self.args)
+
+                save_manifest(self.args, manifest)
+
+                print_output_size(self.args)
+
+            if self.args.verb in ("shell", "boot"):
+                cmdline = run_shell_cmdline(self.args)
+            elif self.args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(run_qemu_cmdline(self.args))
+            else:
+                die("No valid verb was entered.")
+
+            cmdline = [str(x) for x in cmdline]
+            cmdline = " ".join(cmdline)
+
+            self.serial = cmdline
+            self.stack = stack.pop_all()
+
+        return self
+
+    def machine_run(self, command: List[str]) -> None:
+        self.serial.expect("#")
+
+        if self.args.verb == "boot":
+            cmdline = ['sudo', 'systemd-run']
+            cmdline += ["--quiet", "--wait", "--pipe"]
+            cmdline += ["-M", virt_name(self.args)]
+        else:
+            cmdline = ['sudo', 'venv/bin/python', '-m', "mkosi", "ssh"]
```

Comment:
This won't work for a multitude of reasons. There's no reason to call mkosi as a subprocess when we can simply import the relevant function and run it directly. This should be replaced with a call to run_ssh(). run_ssh() itself might have to be modified a little so that we can capture its output.

The new run_ssh() signature should look something like this:

```
def run_ssh(args: MkosiArgs, cmd: Optional[Sequence[str]] = None, stdout: _FILE = None, stderr: _FILE = None) -> CompletedProcess:
```

When using it from the Machine class, we'd call it like this:

```
result  = run_ssh(self.args, command, stdout=PIPE, stderr=PIPE)
```

And then we can access `result.stdout` to get the stdout output of the command and `result.stderr` to get the stderr output of the command.